graphium/fingerprinting/data.py

```diff
@@ -0,0 +1,216 @@
+
+        self.pretrained_models = pretrained_models
+        self.task = task
+        self.benchmark = benchmark
+        self.df_path = df_path
+        self.batch_size = batch_size
+        self.split_type = split_type
+        self.splits_path = splits_path
+        self.split_val = split_val
+        self.split_test = split_test
+        self.data_seed = data_seed
+        self.num_workers = num_workers
+        self.device = device
+        self.mol_cache_dir = mol_cache_dir
+        self.fps_cache_dir = fps_cache_dir
+        if benchmark is not None:
+            # Check if benchmark naming is already implied in config
+            if f"{benchmark}/{task}" not in mol_cache_dir:
+                self.mol_cache_dir = f"{mol_cache_dir}/{benchmark}/{task}"
+            if f"{benchmark}/{task}" not in fps_cache_dir:
+                self.fps_cache_dir = f"{fps_cache_dir}/{benchmark}/{task}"
+
+        self.train_dataset = None
+        self.valid_dataset = None
+        self.test_dataset = None
+
+        self.splits = []
+
+    def prepare_data(self) -> None:  
+        if self.fps_cache_dir is not None and os.path.exists(f"{self.fps_cache_dir}/fps.pt"):
+            self.smiles, self.labels, self.fps_dict = torch.load(f"{self.fps_cache_dir}/fps.pt").values()
+            self.splits = list(self.smiles.keys())
+        
+        else:
+            # Check which splits are needed
+            self.splits = []
+            add_all = self.benchmark is not None or self.splits_path is not None
+            if add_all or self.split_val + self.split_test < 1:
+                self.splits.append("train")
+            if add_all or self.split_val > 0:
+                self.splits.append("valid")
+            if add_all or self.split_test > 0:
+                self.splits.append("test")
+
+            self.data = {
+                split: {
+                    "smiles": [],
+                    "labels": [],
+                    "fps": {},
+                }
+                for split in self.splits
+            }
+
+            for model, layers in self.pretrained_models.items():
+                predictor = PredictorModule.load_pretrained_model(model, device=self.device)
+                predictor.featurization.pop("max_num_atoms", None)
+
+                # Featurization
+                if self.benchmark is None:
+                    assert self.df_path is not None, "df_path must be provided if not using an integrated benchmark"
+
+                    # Add a dummy task column (filled with NaN values) in case no such column is provided
+                    smiles_df = pd.read_csv(self.df_path)
+                    task_cols = [col for col in smiles_df if col.startswith("task_")]
+                    if len(task_cols) == 0:
+                        df_path = ".".join(self.df_path.split(".")[:-1])
+                        smiles_df["task_dummy"] = np.nan
+                        smiles_df.to_csv(f"{df_path}_with_dummy_task_col.csv", index=False)
```

Comment:
Why saving the smiles to CSV?

Reply:
This is to allow users to provide compounds without an associated task column (which is expected by the datamodule). We can save it in the same format as the original file.